doppelganger/doppelganger.go

```diff
@@ -0,0 +1,265 @@
+	if !exists {
+		ds.logger.Warn("Validator not found in Doppelganger state", fields.ValidatorIndex(validatorIndex))
+		return
+	}
+
+	ds.logger.Debug("mark validator as doppelganger safe", fields.ValidatorIndex(validatorIndex))
+	state.remainingEpochs = 0
+}
+
+func (ds *doppelgangerHandler) updateDoppelgangerState(validatorIndices []phase0.ValidatorIndex) {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+
+	// Create a set for current validators
+	currentValidatorSet := make(map[phase0.ValidatorIndex]struct{}, len(validatorIndices))
+	for _, idx := range validatorIndices {
+		currentValidatorSet[idx] = struct{}{}
+	}
+
+	// Add new validators with DefaultRemainingDetectionEpochs
+	for _, idx := range validatorIndices {
+		if _, exists := ds.doppelgangerState[idx]; !exists {
+			ds.doppelgangerState[idx] = &doppelgangerState{
+				remainingEpochs: DefaultRemainingDetectionEpochs,
+			}
+			ds.logger.Debug("Added validator to Doppelganger state", fields.ValidatorIndex(idx))
+		}
+	}
+
+	// Remove validators that are no longer in the current set
+	for idx := range ds.doppelgangerState {
+		if _, exists := currentValidatorSet[idx]; !exists {
+			ds.logger.Debug("Removing validator from Doppelganger state", fields.ValidatorIndex(idx))
+			delete(ds.doppelgangerState, idx)
+		}
+	}
+}
+
+func (ds *doppelgangerHandler) RemoveValidatorState(validatorIndex phase0.ValidatorIndex) {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+
+	if _, exists := ds.doppelgangerState[validatorIndex]; !exists {
+		ds.logger.Warn("Validator not found in Doppelganger state", fields.ValidatorIndex(validatorIndex))
+		return
+	}
+
+	delete(ds.doppelgangerState, validatorIndex)
+	ds.logger.Debug("Removed validator from Doppelganger state", fields.ValidatorIndex(validatorIndex))
+}
+
+func (ds *doppelgangerHandler) StartMonitoring(ctx context.Context) {
+	ds.logger.Info("Doppelganger monitoring started")
+
+	firstRun := true
+	ticker := ds.slotTickerProvider()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.Next():
+			currentSlot := ticker.Slot()
+			currentEpoch := ds.network.Beacon.EstimatedEpochAtSlot(currentSlot)
+			slotsPerEpoch := ds.network.Beacon.SlotsPerEpoch()
+
+			buildStr := fmt.Sprintf("e%v-s%v-#%v", currentEpoch, currentSlot, currentSlot%32+1)
+			ds.logger.Debug("🛠 ticker event", zap.String("epoch_slot_pos", buildStr))
+
+			// Update DG state with self participating validators from validator provider at the current epoch
+			validatorIndices := indicesFromShares(ds.validatorProvider.SelfParticipatingValidators(currentEpoch))
+			ds.updateDoppelgangerState(validatorIndices)
+
+			// Perform liveness checks during the first run or at the last slot of the epoch
+			if (!firstRun && uint64(currentSlot)%slotsPerEpoch != slotsPerEpoch-1) || ds.startEpoch == currentEpoch {
+				continue
+			}
+
+			if firstRun {
+				ds.startEpoch = currentEpoch
+				firstRun = false
+			}
+
+			ds.checkLiveness(ctx, currentEpoch-1)
```

Comment:
IMO we should set a deadline for this ctx until the end of the slot with -200ms or more for some safety margin